File: clients/rn/src/network/AuthClient.ts
/**
 * AuthClient — HTTP client for the account/auth API.
 *
 * Used by LoginScreen after pairing to create or authenticate a user
 * account on the household server. Mirrors the KMP AuthClient.
 *
 * Uses standard fetch() — no external HTTP library needed.
 */

export interface AuthResult {
  token: string;
  userId: string;
  username: string;
  role: string;
}

export interface ServerStatus {
  hasUsers: boolean;
  openRegistration: boolean;
}

function normalizeUrl(url: string): string {
  let u = url.trim().replace(/\/+$/, '');
  if (!/^https?:\/\//i.test(u)) {
    u = 'http://' + u;
  }
  return u;
}

/**
 * Query GET /api/auth/status to determine whether the server has
 * existing user accounts (and whether open registration is allowed).
 */
export async function checkStatus(serverUrl: string): Promise<ServerStatus> {
  const url = normalizeUrl(serverUrl);
  const res = await fetch(`${url}/api/auth/status`);
  if (!res.ok) {
    throw new Error(`Server status check failed: ${res.status}`);
  }
  const body = await res.json();
  return {
    hasUsers: body.hasUsers ?? body.has_users ?? false,
    openRegistration: body.openRegistration ?? body.open_registration ?? true,
  };
}

/**
 * POST /api/auth/login — authenticate with username + password.
 */
export async function login(
  serverUrl: string,
  username: string,
  password: string,
): Promise<AuthResult> {
  const url = normalizeUrl(serverUrl);
  const res = await fetch(`${url}/api/auth/login`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ username, password }),
  });
  if (!res.ok) {
    if (res.status === 401) {
      throw new Error('Invalid username or password.');
    }
    const text = await res.text().catch(() => '');
    throw new Error(
      text.includes('username') || text.includes('password')
        ? 'Invalid username or password.'
        : `Login failed: ${res.status}`,
    );
  }
  const body = await res.json();
  return {
    token: body.token,
    userId: body.userId ?? body.user_id,
    username: body.username,
    role: body.role ?? 'user',
  };
}

/**
 * POST /api/auth/register — create a new account.
 */
export async function register(
  serverUrl: string,
  username: string,
  password: string,
  displayName?: string,
): Promise<AuthResult> {
  const url = normalizeUrl(serverUrl);
  const res = await fetch(`${url}/api/auth/register`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      username,
      password,
      display_name: displayName ?? username,
    }),
  });
  if (!res.ok) {
    if (res.status === 409) {
      throw new Error('Username already taken.');
    }
    const text = await res.text().catch(() => '');
    throw new Error(
      text.includes('taken') || text.includes('exists')
        ? 'Username already taken.'
        : `Registration failed: ${res.status}`,
    );
  }
  const body = await res.json();
  return {
    token: body.token,
    userId: body.userId ?? body.user_id,
    username: body.username,
    role: body.role ?? 'user',
  };
}

/**
 * POST /api/auth/link-device — bind a pairing device token to the
 * authenticated user account.
 */
export async function linkDevice(
  serverUrl: string,
  authToken: string,
  deviceToken: string,
): Promise<boolean> {
  const url = normalizeUrl(serverUrl);
  try {
    const res = await fetch(`${url}/api/auth/link-device`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${authToken}`,
      },
      body: JSON.stringify({ deviceToken }),
    });
    return res.ok;
  } catch {
    // Link failure is non-fatal — user is still authenticated
    return false;
  }
}

File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/network/AuthClient.kt
package org.wyrdsekai.app.network

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

@Serializable
data class AuthResponse(
    val token: String,
    @kotlinx.serialization.SerialName("userId")
    val user_id: String,
    val username: String,
    val role: String = "user",
)

@Serializable
data class UserInfo(
    val user_id: String,
    val username: String,
    val display_name: String?,
)

@Serializable
data class ServerStatus(
    @kotlinx.serialization.SerialName("hasUsers")
    val has_users: Boolean,
    @kotlinx.serialization.SerialName("openRegistration")
    val open_registration: Boolean = true,
)

@Serializable
data class LinkDeviceResponse(
    val linked: Boolean = true,
)

/** Auto-prepend http:// if no scheme given. */
internal fun normalizeHttpUrl(url: String): String {
    val trimmed = url.trim().trimEnd('/')
    return if (trimmed.startsWith("http://") || trimmed.startsWith("https://")) trimmed
    else "http://$trimmed"
}

class AuthClient(baseUrl: String) {
    private val normalizedUrl = normalizeHttpUrl(baseUrl)
    private val http = HttpClient {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
    }

    suspend fun register(username: String, password: String, displayName: String): Result<AuthResponse> =
        runCatching {
            val response = http.post("$normalizedUrl/api/auth/register") {
                contentType(ContentType.Application.Json)
                setBody(mapOf(
                    "username" to username,
                    "password" to password,
                    "display_name" to displayName,
                ))
            }
            if (!response.status.isSuccess()) {
                val body = response.bodyAsText()
                throw IllegalStateException(
                    try { Json.parseToJsonElement(body).jsonObject["error"]?.jsonPrimitive?.content ?: body }
                    catch (_: Exception) { body }
                )
            }
            response.body()
        }

    suspend fun login(username: String, password: String): Result<AuthResponse> =
        runCatching {
            val response = http.post("$normalizedUrl/api/auth/login") {
                contentType(ContentType.Application.Json)
                setBody(mapOf(
                    "username" to username,
                    "password" to password,
                ))
            }
            if (!response.status.isSuccess()) {
                val body = response.bodyAsText()
                throw IllegalStateException(
                    try { Json.parseToJsonElement(body).jsonObject["error"]?.jsonPrimitive?.content ?: body }
                    catch (_: Exception) { body }
                )
            }
            response.body()
        }

    suspend fun me(token: String): Result<UserInfo> =
        runCatching {
            http.get("$normalizedUrl/api/auth/me") {
                parameter("token", token)
            }.body()
        }

    suspend fun checkStatus(): Result<ServerStatus> =
        runCatching {
            http.get("$normalizedUrl/api/auth/status").body()
        }

    suspend fun linkDevice(authToken: String, deviceToken: String): Result<LinkDeviceResponse> =
        runCatching {
            http.post("$normalizedUrl/api/auth/link-device") {
                contentType(ContentType.Application.Json)
                header("Authorization", "Bearer $authToken")
                setBody(mapOf("deviceToken" to deviceToken))
            }.body()
        }

    fun close() {
        http.close()
    }
}
